cpp/benchmarks/pairwise_linestring_distance.cu

```diff
@@ -0,0 +1,170 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <benchmarks/fixture/rmm_pool_raii.hpp>
+#include <nvbench/nvbench.cuh>
+
+#include <cuspatial/experimental/linestring_distance.cuh>
+#include <cuspatial/experimental/type_utils.hpp>
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <memory>
+
+namespace cuspatial {
+
+/**
+ * @brief Helper to generate linestrings used for benchmarks.
+ *
+ * The generator adopts a walking algorithm. The ith point is computed by
+ * walking (cos(i) * segment_length, sin(i) * segment_length) from the `i-1`
+ * point. The initial point of the linestring is at `(init_xy, init_xy)`.
+ *
+ * The number of line segments per linestring is constrolled by
+ * `num_segment_per_string`.
+ *
+ * Since the outreach upper bound of the linestring group is
+ * `(init_xy + num_strings * num_segments_per_string * segment_length)`,
+ * user may control the locality of the linestring group via these four
+ * arguments.
+ *
+ * @tparam T The floating point type for the coordinates
+ * @param num_strings Total number of linestrings
+ * @param num_segments_per_string Number of line segments per linestring
+ * @param segment_length Length of each segment, or stride of walk
+ * @param init_xy The initial coordinate to start the walk
+ * @param stream The CUDA stream to use for device memory operations and kernel launches
+ * @return A tuple of x and y coordinates of points and offsets to which the first point
+ * of each linestring starts.
+ *
+ */
+template <typename T>
+std::tuple<std::unique_ptr<cudf::column>,
+           std::unique_ptr<cudf::column>,
+           rmm::device_uvector<cudf::size_type>>
+generate_linestring(cudf::size_type num_strings,
+                    cudf::size_type num_segments_per_string,
+                    T segment_length,
+                    T init_xy,
+                    rmm::cuda_stream_view stream)
+{
+  cudf::size_type num_points = num_strings * (num_segments_per_string + 1);
+  rmm::device_uvector<cudf::size_type> offsets(num_points, stream);
+  thrust::transform(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator(static_cast<cudf::size_type>(0)),
+    thrust::make_counting_iterator(static_cast<cudf::size_type>(num_points)),
+    offsets.begin(),
+    [num_segments_per_string] __device__(auto i) { return i * num_segments_per_string; });
+  auto points_x = cudf::make_fixed_width_column(
```

Comment:
Hmm, now that you mention, you are right about that. The current code is adapted from an earlier version that uses the cudf column API. With header only API in place, we can bypass using columns in all.

Reply:
Well actually, since there's an `exclusive_scan` call I'm not sure if this is possible. My knowledge of thrust iterators tells me we can't make an `exclusive_scan` iterator.